File: Medium/767. Reorganize String/767. Reorganize String.py
class Solution:
    def reorganizeString(self, s: str) -> str:
        char_count = {}
        for char in s:
            char_count[char] = char_count.get(char, 0) + 1

        max_heap = [(-count, char) for char, count in char_count.items()]
        heapq.heapify(max_heap)

        result = []
        prev_count, prev_char = 0, ""

        while max_heap:
            count, char = heapq.heappop(max_heap)
            result.append(char)

            if prev_count < 0:
                heapq.heappush(max_heap, (prev_count, prev_char))

            count += 1
            prev_count, prev_char = count, char

        if len(result) == len(s):
            return "".join(result)
        else:
            return ""

File: Medium/767. Reorganize String/767. Reorganize String.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn reorganize_string(s: String) -> String {
        let mut char_count = vec![0; 26];
        for c in s.chars() {
            char_count[(c as u8 - b'a') as usize] += 1;
        }

        let mut max_heap = BinaryHeap::new();
        for (i, &count) in char_count.iter().enumerate() {
            if count > 0 {
                max_heap.push((count, (i as u8 + b'a') as char));
            }
        }

        let mut result = Vec::new();
        while let Some((count1, char1)) = max_heap.pop() {
            if let Some((count2, char2)) = max_heap.pop() {
                result.push(char1);
                result.push(char2);
                if count1 > 1 {
                    max_heap.push((count1 - 1, char1));
                }
                if count2 > 1 {
                    max_heap.push((count2 - 1, char2));
                }
            } else {
                if count1 > 1 {
                    return String::new(); // Not possible to rearrange
                } else {
                    result.push(char1);
                }
            }
        }

        result.into_iter().collect()
    }
}
